// node-stdlib-browser/cjs/proxy/constants.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var O_RDONLY = 0;
var O_WRONLY = 1;
var O_RDWR = 2;
var S_IFMT = 61440;
var S_IFREG = 32768;
var S_IFDIR = 16384;
var S_IFCHR = 8192;
var S_IFBLK = 24576;
var S_IFIFO = 4096;
var S_IFLNK = 40960;
var S_IFSOCK = 49152;
var O_CREAT = 512;
var O_EXCL = 2048;
var O_NOCTTY = 131072;
var O_TRUNC = 1024;
var O_APPEND = 8;
var O_DIRECTORY = 1048576;
var O_NOFOLLOW = 256;
var O_SYNC = 128;
var O_SYMLINK = 2097152;
var O_NONBLOCK = 4;
var S_IRWXU = 448;
var S_IRUSR = 256;
var S_IWUSR = 128;
var S_IXUSR = 64;
var S_IRWXG = 56;
var S_IRGRP = 32;
var S_IWGRP = 16;
var S_IXGRP = 8;
var S_IRWXO = 7;
var S_IROTH = 4;
var S_IWOTH = 2;
var S_IXOTH = 1;
var E2BIG = 7;
var EACCES = 13;
var EADDRINUSE = 48;
var EADDRNOTAVAIL = 49;
var EAFNOSUPPORT = 47;
var EAGAIN = 35;
var EALREADY = 37;
var EBADF = 9;
var EBADMSG = 94;
var EBUSY = 16;
var ECANCELED = 89;
var ECHILD = 10;
var ECONNABORTED = 53;
var ECONNREFUSED = 61;
var ECONNRESET = 54;
var EDEADLK = 11;
var EDESTADDRREQ = 39;
var EDOM = 33;
var EDQUOT = 69;
var EEXIST = 17;
var EFAULT = 14;
var EFBIG = 27;
var EHOSTUNREACH = 65;
var EIDRM = 90;
var EILSEQ = 92;
var EINPROGRESS = 36;
var EINTR = 4;
var EINVAL = 22;
var EIO = 5;
var EISCONN = 56;
var EISDIR = 21;
var ELOOP = 62;
var EMFILE = 24;
var EMLINK = 31;
var EMSGSIZE = 40;
var EMULTIHOP = 95;
var ENAMETOOLONG = 63;
var ENETDOWN = 50;
var ENETRESET = 52;
var ENETUNREACH = 51;
var ENFILE = 23;
var ENOBUFS = 55;
var ENODATA = 96;
var ENODEV = 19;
var ENOENT = 2;
var ENOEXEC = 8;
var ENOLCK = 77;
var ENOLINK = 97;
var ENOMEM = 12;
var ENOMSG = 91;
var ENOPROTOOPT = 42;
var ENOSPC = 28;
var ENOSR = 98;
var ENOSTR = 99;
var ENOSYS = 78;
var ENOTCONN = 57;
var ENOTDIR = 20;
var ENOTEMPTY = 66;
var ENOTSOCK = 38;
var ENOTSUP = 45;
var ENOTTY = 25;
var ENXIO = 6;
var EOPNOTSUPP = 102;
var EOVERFLOW = 84;
var EPERM = 1;
var EPIPE = 32;
var EPROTO = 100;
var EPROTONOSUPPORT = 43;
var EPROTOTYPE = 41;
var ERANGE = 34;
var EROFS = 30;
var ESPIPE = 29;
var ESRCH = 3;
var ESTALE = 70;
var ETIME = 101;
var ETIMEDOUT = 60;
var ETXTBSY = 26;
var EWOULDBLOCK = 35;
var EXDEV = 18;
var SIGHUP = 1;
var SIGINT = 2;
var SIGQUIT = 3;
var SIGILL = 4;
var SIGTRAP = 5;
var SIGABRT = 6;
var SIGIOT = 6;
var SIGBUS = 10;
var SIGFPE = 8;
var SIGKILL = 9;
var SIGUSR1 = 30;
var SIGSEGV = 11;
var SIGUSR2 = 31;
var SIGPIPE = 13;
var SIGALRM = 14;
var SIGTERM = 15;
var SIGCHLD = 20;
var SIGCONT = 19;
var SIGSTOP = 17;
var SIGTSTP = 18;
var SIGTTIN = 21;
var SIGTTOU = 22;
var SIGURG = 16;
var SIGXCPU = 24;
var SIGXFSZ = 25;
var SIGVTALRM = 26;
var SIGPROF = 27;
var SIGWINCH = 28;
var SIGIO = 23;
var SIGSYS = 12;
var SSL_OP_ALL = 2147486719;
var SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 262144;
var SSL_OP_CIPHER_SERVER_PREFERENCE = 4194304;
var SSL_OP_CISCO_ANYCONNECT = 32768;
var SSL_OP_COOKIE_EXCHANGE = 8192;
var SSL_OP_CRYPTOPRO_TLSEXT_BUG = 2147483648;
var SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS = 2048;
var SSL_OP_EPHEMERAL_RSA = 0;
var SSL_OP_LEGACY_SERVER_CONNECT = 4;
var SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER = 32;
var SSL_OP_MICROSOFT_SESS_ID_BUG = 1;
var SSL_OP_MSIE_SSLV2_RSA_PADDING = 0;
var SSL_OP_NETSCAPE_CA_DN_BUG = 536870912;
var SSL_OP_NETSCAPE_CHALLENGE_BUG = 2;
var SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG = 1073741824;
var SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG = 8;
var SSL_OP_NO_COMPRESSION = 131072;
var SSL_OP_NO_QUERY_MTU = 4096;
var SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = 65536;
var SSL_OP_NO_SSLv2 = 16777216;
var SSL_OP_NO_SSLv3 = 33554432;
var SSL_OP_NO_TICKET = 16384;
var SSL_OP_NO_TLSv1 = 67108864;
var SSL_OP_NO_TLSv1_1 = 268435456;
var SSL_OP_NO_TLSv1_2 = 134217728;
var SSL_OP_PKCS1_CHECK_1 = 0;
var SSL_OP_PKCS1_CHECK_2 = 0;
var SSL_OP_SINGLE_DH_USE = 1048576;
var SSL_OP_SINGLE_ECDH_USE = 524288;
var SSL_OP_SSLEAY_080_CLIENT_DH_BUG = 128;
var SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG = 0;
var SSL_OP_TLS_BLOCK_PADDING_BUG = 512;
var SSL_OP_TLS_D5_BUG = 256;
var SSL_OP_TLS_ROLLBACK_BUG = 8388608;
var ENGINE_METHOD_DSA = 2;
var ENGINE_METHOD_DH = 4;
var ENGINE_METHOD_RAND = 8;
var ENGINE_METHOD_ECDH = 16;
var ENGINE_METHOD_ECDSA = 32;
var ENGINE_METHOD_CIPHERS = 64;
var ENGINE_METHOD_DIGESTS = 128;
var ENGINE_METHOD_STORE = 256;
var ENGINE_METHOD_PKEY_METHS = 512;
var ENGINE_METHOD_PKEY_ASN1_METHS = 1024;
var ENGINE_METHOD_ALL = 65535;
var ENGINE_METHOD_NONE = 0;
var DH_CHECK_P_NOT_SAFE_PRIME = 2;
var DH_CHECK_P_NOT_PRIME = 1;
var DH_UNABLE_TO_CHECK_GENERATOR = 4;
var DH_NOT_SUITABLE_GENERATOR = 8;
var NPN_ENABLED = 1;
var RSA_PKCS1_PADDING = 1;
var RSA_SSLV23_PADDING = 2;
var RSA_NO_PADDING = 3;
var RSA_PKCS1_OAEP_PADDING = 4;
var RSA_X931_PADDING = 5;
var RSA_PKCS1_PSS_PADDING = 6;
var POINT_CONVERSION_COMPRESSED = 2;
var POINT_CONVERSION_UNCOMPRESSED = 4;
var POINT_CONVERSION_HYBRID = 6;
var F_OK = 0;
var R_OK = 4;
var W_OK = 2;
var X_OK = 1;
var UV_UDP_REUSEADDR = 4;
var api = {
  O_RDONLY: O_RDONLY,
  O_WRONLY: O_WRONLY,
  O_RDWR: O_RDWR,
  S_IFMT: S_IFMT,
  S_IFREG: S_IFREG,
  S_IFDIR: S_IFDIR,
  S_IFCHR: S_IFCHR,
  S_IFBLK: S_IFBLK,
  S_IFIFO: S_IFIFO,
  S_IFLNK: S_IFLNK,
  S_IFSOCK: S_IFSOCK,
  O_CREAT: O_CREAT,
  O_EXCL: O_EXCL,
  O_NOCTTY: O_NOCTTY,
  O_TRUNC: O_TRUNC,
  O_APPEND: O_APPEND,
  O_DIRECTORY: O_DIRECTORY,
  O_NOFOLLOW: O_NOFOLLOW,
  O_SYNC: O_SYNC,
  O_SYMLINK: O_SYMLINK,
  O_NONBLOCK: O_NONBLOCK,
  S_IRWXU: S_IRWXU,
  S_IRUSR: S_IRUSR,
  S_IWUSR: S_IWUSR,
  S_IXUSR: S_IXUSR,
  S_IRWXG: S_IRWXG,
  S_IRGRP: S_IRGRP,
  S_IWGRP: S_IWGRP,
  S_IXGRP: S_IXGRP,
  S_IRWXO: S_IRWXO,
  S_IROTH: S_IROTH,
  S_IWOTH: S_IWOTH,
  S_IXOTH: S_IXOTH,
  E2BIG: E2BIG,
  EACCES: EACCES,
  EADDRINUSE: EADDRINUSE,
  EADDRNOTAVAIL: EADDRNOTAVAIL,
  EAFNOSUPPORT: EAFNOSUPPORT,
  EAGAIN: EAGAIN,
  EALREADY: EALREADY,
  EBADF: EBADF,
  EBADMSG: EBADMSG,
  EBUSY: EBUSY,
  ECANCELED: ECANCELED,
  ECHILD: ECHILD,
  ECONNABORTED: ECONNABORTED,
  ECONNREFUSED: ECONNREFUSED,
  ECONNRESET: ECONNRESET,
  EDEADLK: EDEADLK,
  EDESTADDRREQ: EDESTADDRREQ,
  EDOM: EDOM,
  EDQUOT: EDQUOT,
  EEXIST: EEXIST,
  EFAULT: EFAULT,
  EFBIG: EFBIG,
  EHOSTUNREACH: EHOSTUNREACH,
  EIDRM: EIDRM,
  EILSEQ: EILSEQ,
  EINPROGRESS: EINPROGRESS,
  EINTR: EINTR,
  EINVAL: EINVAL,
  EIO: EIO,
  EISCONN: EISCONN,
  EISDIR: EISDIR,
  ELOOP: ELOOP,
  EMFILE: EMFILE,
  EMLINK: EMLINK,
  EMSGSIZE: EMSGSIZE,
  EMULTIHOP: EMULTIHOP,
  ENAMETOOLONG: ENAMETOOLONG,
  ENETDOWN: ENETDOWN,
  ENETRESET: ENETRESET,
  ENETUNREACH: ENETUNREACH,
  ENFILE: ENFILE,
  ENOBUFS: ENOBUFS,
  ENODATA: ENODATA,
  ENODEV: ENODEV,
  ENOENT: ENOENT,
  ENOEXEC: ENOEXEC,
  ENOLCK: ENOLCK,
  ENOLINK: ENOLINK,
  ENOMEM: ENOMEM,
  ENOMSG: ENOMSG,
  ENOPROTOOPT: ENOPROTOOPT,
  ENOSPC: ENOSPC,
  ENOSR: ENOSR,
  ENOSTR: ENOSTR,
  ENOSYS: ENOSYS,
  ENOTCONN: ENOTCONN,
  ENOTDIR: ENOTDIR,
  ENOTEMPTY: ENOTEMPTY,
  ENOTSOCK: ENOTSOCK,
  ENOTSUP: ENOTSUP,
  ENOTTY: ENOTTY,
  ENXIO: ENXIO,
  EOPNOTSUPP: EOPNOTSUPP,
  EOVERFLOW: EOVERFLOW,
  EPERM: EPERM,
  EPIPE: EPIPE,
  EPROTO: EPROTO,
  EPROTONOSUPPORT: EPROTONOSUPPORT,
  EPROTOTYPE: EPROTOTYPE,
  ERANGE: ERANGE,
  EROFS: EROFS,
  ESPIPE: ESPIPE,
  ESRCH: ESRCH,
  ESTALE: ESTALE,
  ETIME: ETIME,
  ETIMEDOUT: ETIMEDOUT,
  ETXTBSY: ETXTBSY,
  EWOULDBLOCK: EWOULDBLOCK,
  EXDEV: EXDEV,
  SIGHUP: SIGHUP,
  SIGINT: SIGINT,
  SIGQUIT: SIGQUIT,
  SIGILL: SIGILL,
  SIGTRAP: SIGTRAP,
  SIGABRT: SIGABRT,
  SIGIOT: SIGIOT,
  SIGBUS: SIGBUS,
  SIGFPE: SIGFPE,
  SIGKILL: SIGKILL,
  SIGUSR1: SIGUSR1,
  SIGSEGV: SIGSEGV,
  SIGUSR2: SIGUSR2,
  SIGPIPE: SIGPIPE,
  SIGALRM: SIGALRM,
  SIGTERM: SIGTERM,
  SIGCHLD: SIGCHLD,
  SIGCONT: SIGCONT,
  SIGSTOP: SIGSTOP,
  SIGTSTP: SIGTSTP,
  SIGTTIN: SIGTTIN,
  SIGTTOU: SIGTTOU,
  SIGURG: SIGURG,
  SIGXCPU: SIGXCPU,
  SIGXFSZ: SIGXFSZ,
  SIGVTALRM: SIGVTALRM,
  SIGPROF: SIGPROF,
  SIGWINCH: SIGWINCH,
  SIGIO: SIGIO,
  SIGSYS: SIGSYS,
  SSL_OP_ALL: SSL_OP_ALL,
  SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION: SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION,
  SSL_OP_CIPHER_SERVER_PREFERENCE: SSL_OP_CIPHER_SERVER_PREFERENCE,
  SSL_OP_CISCO_ANYCONNECT: SSL_OP_CISCO_ANYCONNECT,
  SSL_OP_COOKIE_EXCHANGE: SSL_OP_COOKIE_EXCHANGE,
  SSL_OP_CRYPTOPRO_TLSEXT_BUG: SSL_OP_CRYPTOPRO_TLSEXT_BUG,
  SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS: SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS,
  SSL_OP_EPHEMERAL_RSA: SSL_OP_EPHEMERAL_RSA,
  SSL_OP_LEGACY_SERVER_CONNECT: SSL_OP_LEGACY_SERVER_CONNECT,
  SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER: SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER,
  SSL_OP_MICROSOFT_SESS_ID_BUG: SSL_OP_MICROSOFT_SESS_ID_BUG,
  SSL_OP_MSIE_SSLV2_RSA_PADDING: SSL_OP_MSIE_SSLV2_RSA_PADDING,
  SSL_OP_NETSCAPE_CA_DN_BUG: SSL_OP_NETSCAPE_CA_DN_BUG,
  SSL_OP_NETSCAPE_CHALLENGE_BUG: SSL_OP_NETSCAPE_CHALLENGE_BUG,
  SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG: SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG,
  SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG: SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG,
  SSL_OP_NO_COMPRESSION: SSL_OP_NO_COMPRESSION,
  SSL_OP_NO_QUERY_MTU: SSL_OP_NO_QUERY_MTU,
  SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION: SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION,
  SSL_OP_NO_SSLv2: SSL_OP_NO_SSLv2,
  SSL_OP_NO_SSLv3: SSL_OP_NO_SSLv3,
  SSL_OP_NO_TICKET: SSL_OP_NO_TICKET,
  SSL_OP_NO_TLSv1: SSL_OP_NO_TLSv1,
  SSL_OP_NO_TLSv1_1: SSL_OP_NO_TLSv1_1,
  SSL_OP_NO_TLSv1_2: SSL_OP_NO_TLSv1_2,
  SSL_OP_PKCS1_CHECK_1: SSL_OP_PKCS1_CHECK_1,
  SSL_OP_PKCS1_CHECK_2: SSL_OP_PKCS1_CHECK_2,
  SSL_OP_SINGLE_DH_USE: SSL_OP_SINGLE_DH_USE,
  SSL_OP_SINGLE_ECDH_USE: SSL_OP_SINGLE_ECDH_USE,
  SSL_OP_SSLEAY_080_CLIENT_DH_BUG: SSL_OP_SSLEAY_080_CLIENT_DH_BUG,
  SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG: SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG,
  SSL_OP_TLS_BLOCK_PADDING_BUG: SSL_OP_TLS_BLOCK_PADDING_BUG,
  SSL_OP_TLS_D5_BUG: SSL_OP_TLS_D5_BUG,
  SSL_OP_TLS_ROLLBACK_BUG: SSL_OP_TLS_ROLLBACK_BUG,
  ENGINE_METHOD_DSA: ENGINE_METHOD_DSA,
  ENGINE_METHOD_DH: ENGINE_METHOD_DH,
  ENGINE_METHOD_RAND: ENGINE_METHOD_RAND,
  ENGINE_METHOD_ECDH: ENGINE_METHOD_ECDH,
  ENGINE_METHOD_ECDSA: ENGINE_METHOD_ECDSA,
  ENGINE_METHOD_CIPHERS: ENGINE_METHOD_CIPHERS,
  ENGINE_METHOD_DIGESTS: ENGINE_METHOD_DIGESTS,
  ENGINE_METHOD_STORE: ENGINE_METHOD_STORE,
  ENGINE_METHOD_PKEY_METHS: ENGINE_METHOD_PKEY_METHS,
  ENGINE_METHOD_PKEY_ASN1_METHS: ENGINE_METHOD_PKEY_ASN1_METHS,
  ENGINE_METHOD_ALL: ENGINE_METHOD_ALL,
  ENGINE_METHOD_NONE: ENGINE_METHOD_NONE,
  DH_CHECK_P_NOT_SAFE_PRIME: DH_CHECK_P_NOT_SAFE_PRIME,
  DH_CHECK_P_NOT_PRIME: DH_CHECK_P_NOT_PRIME,
  DH_UNABLE_TO_CHECK_GENERATOR: DH_UNABLE_TO_CHECK_GENERATOR,
  DH_NOT_SUITABLE_GENERATOR: DH_NOT_SUITABLE_GENERATOR,
  NPN_ENABLED: NPN_ENABLED,
  RSA_PKCS1_PADDING: RSA_PKCS1_PADDING,
  RSA_SSLV23_PADDING: RSA_SSLV23_PADDING,
  RSA_NO_PADDING: RSA_NO_PADDING,
  RSA_PKCS1_OAEP_PADDING: RSA_PKCS1_OAEP_PADDING,
  RSA_X931_PADDING: RSA_X931_PADDING,
  RSA_PKCS1_PSS_PADDING: RSA_PKCS1_PSS_PADDING,
  POINT_CONVERSION_COMPRESSED: POINT_CONVERSION_COMPRESSED,
  POINT_CONVERSION_UNCOMPRESSED: POINT_CONVERSION_UNCOMPRESSED,
  POINT_CONVERSION_HYBRID: POINT_CONVERSION_HYBRID,
  F_OK: F_OK,
  R_OK: R_OK,
  W_OK: W_OK,
  X_OK: X_OK,
  UV_UDP_REUSEADDR: UV_UDP_REUSEADDR
};

exports.DH_CHECK_P_NOT_PRIME = DH_CHECK_P_NOT_PRIME;
exports.DH_CHECK_P_NOT_SAFE_PRIME = DH_CHECK_P_NOT_SAFE_PRIME;
exports.DH_NOT_SUITABLE_GENERATOR = DH_NOT_SUITABLE_GENERATOR;
exports.DH_UNABLE_TO_CHECK_GENERATOR = DH_UNABLE_TO_CHECK_GENERATOR;
exports.E2BIG = E2BIG;
exports.EACCES = EACCES;
exports.EADDRINUSE = EADDRINUSE;
exports.EADDRNOTAVAIL = EADDRNOTAVAIL;
exports.EAFNOSUPPORT = EAFNOSUPPORT;
exports.EAGAIN = EAGAIN;
exports.EALREADY = EALREADY;
exports.EBADF = EBADF;
exports.EBADMSG = EBADMSG;
exports.EBUSY = EBUSY;
exports.ECANCELED = ECANCELED;
exports.ECHILD = ECHILD;
exports.ECONNABORTED = ECONNABORTED;
exports.ECONNREFUSED = ECONNREFUSED;
exports.ECONNRESET = ECONNRESET;
exports.EDEADLK = EDEADLK;
exports.EDESTADDRREQ = EDESTADDRREQ;
exports.EDOM = EDOM;
exports.EDQUOT = EDQUOT;
exports.EEXIST = EEXIST;
exports.EFAULT = EFAULT;
exports.EFBIG = EFBIG;
exports.EHOSTUNREACH = EHOSTUNREACH;
exports.EIDRM = EIDRM;
exports.EILSEQ = EILSEQ;
exports.EINPROGRESS = EINPROGRESS;
exports.EINTR = EINTR;
exports.EINVAL = EINVAL;
exports.EIO = EIO;
exports.EISCONN = EISCONN;
exports.EISDIR = EISDIR;
exports.ELOOP = ELOOP;
exports.EMFILE = EMFILE;
exports.EMLINK = EMLINK;
exports.EMSGSIZE = EMSGSIZE;
exports.EMULTIHOP = EMULTIHOP;
exports.ENAMETOOLONG = ENAMETOOLONG;
exports.ENETDOWN = ENETDOWN;
exports.ENETRESET = ENETRESET;
exports.ENETUNREACH = ENETUNREACH;
exports.ENFILE = ENFILE;
exports.ENGINE_METHOD_ALL = ENGINE_METHOD_ALL;
exports.ENGINE_METHOD_CIPHERS = ENGINE_METHOD_CIPHERS;
exports.ENGINE_METHOD_DH = ENGINE_METHOD_DH;
exports.ENGINE_METHOD_DIGESTS = ENGINE_METHOD_DIGESTS;
exports.ENGINE_METHOD_DSA = ENGINE_METHOD_DSA;
exports.ENGINE_METHOD_ECDH = ENGINE_METHOD_ECDH;
exports.ENGINE_METHOD_ECDSA = ENGINE_METHOD_ECDSA;
exports.ENGINE_METHOD_NONE = ENGINE_METHOD_NONE;
exports.ENGINE_METHOD_PKEY_ASN1_METHS = ENGINE_METHOD_PKEY_ASN1_METHS;
exports.ENGINE_METHOD_PKEY_METHS = ENGINE_METHOD_PKEY_METHS;
exports.ENGINE_METHOD_RAND = ENGINE_METHOD_RAND;
exports.ENGINE_METHOD_STORE = ENGINE_METHOD_STORE;
exports.ENOBUFS = ENOBUFS;
exports.ENODATA = ENODATA;
exports.ENODEV = ENODEV;
exports.ENOENT = ENOENT;
exports.ENOEXEC = ENOEXEC;
exports.ENOLCK = ENOLCK;
exports.ENOLINK = ENOLINK;
exports.ENOMEM = ENOMEM;
exports.ENOMSG = ENOMSG;
exports.ENOPROTOOPT = ENOPROTOOPT;
exports.ENOSPC = ENOSPC;
exports.ENOSR = ENOSR;
exports.ENOSTR = ENOSTR;
exports.ENOSYS = ENOSYS;
exports.ENOTCONN = ENOTCONN;
exports.ENOTDIR = ENOTDIR;
exports.ENOTEMPTY = ENOTEMPTY;
exports.ENOTSOCK = ENOTSOCK;
exports.ENOTSUP = ENOTSUP;
exports.ENOTTY = ENOTTY;
exports.ENXIO = ENXIO;
exports.EOPNOTSUPP = EOPNOTSUPP;
exports.EOVERFLOW = EOVERFLOW;
exports.EPERM = EPERM;
exports.EPIPE = EPIPE;
exports.EPROTO = EPROTO;
exports.EPROTONOSUPPORT = EPROTONOSUPPORT;
exports.EPROTOTYPE = EPROTOTYPE;
exports.ERANGE = ERANGE;
exports.EROFS = EROFS;
exports.ESPIPE = ESPIPE;
exports.ESRCH = ESRCH;
exports.ESTALE = ESTALE;
exports.ETIME = ETIME;
exports.ETIMEDOUT = ETIMEDOUT;
exports.ETXTBSY = ETXTBSY;
exports.EWOULDBLOCK = EWOULDBLOCK;
exports.EXDEV = EXDEV;
exports.F_OK = F_OK;
exports.NPN_ENABLED = NPN_ENABLED;
exports.O_APPEND = O_APPEND;
exports.O_CREAT = O_CREAT;
exports.O_DIRECTORY = O_DIRECTORY;
exports.O_EXCL = O_EXCL;
exports.O_NOCTTY = O_NOCTTY;
exports.O_NOFOLLOW = O_NOFOLLOW;
exports.O_NONBLOCK = O_NONBLOCK;
exports.O_RDONLY = O_RDONLY;
exports.O_RDWR = O_RDWR;
exports.O_SYMLINK = O_SYMLINK;
exports.O_SYNC = O_SYNC;
exports.O_TRUNC = O_TRUNC;
exports.O_WRONLY = O_WRONLY;
exports.POINT_CONVERSION_COMPRESSED = POINT_CONVERSION_COMPRESSED;
exports.POINT_CONVERSION_HYBRID = POINT_CONVERSION_HYBRID;
exports.POINT_CONVERSION_UNCOMPRESSED = POINT_CONVERSION_UNCOMPRESSED;
exports.RSA_NO_PADDING = RSA_NO_PADDING;
exports.RSA_PKCS1_OAEP_PADDING = RSA_PKCS1_OAEP_PADDING;
exports.RSA_PKCS1_PADDING = RSA_PKCS1_PADDING;
exports.RSA_PKCS1_PSS_PADDING = RSA_PKCS1_PSS_PADDING;
exports.RSA_SSLV23_PADDING = RSA_SSLV23_PADDING;
exports.RSA_X931_PADDING = RSA_X931_PADDING;
exports.R_OK = R_OK;
exports.SIGABRT = SIGABRT;
exports.SIGALRM = SIGALRM;
exports.SIGBUS = SIGBUS;
exports.SIGCHLD = SIGCHLD;
exports.SIGCONT = SIGCONT;
exports.SIGFPE = SIGFPE;
exports.SIGHUP = SIGHUP;
exports.SIGILL = SIGILL;
exports.SIGINT = SIGINT;
exports.SIGIO = SIGIO;
exports.SIGIOT = SIGIOT;
exports.SIGKILL = SIGKILL;
exports.SIGPIPE = SIGPIPE;
exports.SIGPROF = SIGPROF;
exports.SIGQUIT = SIGQUIT;
exports.SIGSEGV = SIGSEGV;
exports.SIGSTOP = SIGSTOP;
exports.SIGSYS = SIGSYS;
exports.SIGTERM = SIGTERM;
exports.SIGTRAP = SIGTRAP;
exports.SIGTSTP = SIGTSTP;
exports.SIGTTIN = SIGTTIN;
exports.SIGTTOU = SIGTTOU;
exports.SIGURG = SIGURG;
exports.SIGUSR1 = SIGUSR1;
exports.SIGUSR2 = SIGUSR2;
exports.SIGVTALRM = SIGVTALRM;
exports.SIGWINCH = SIGWINCH;
exports.SIGXCPU = SIGXCPU;
exports.SIGXFSZ = SIGXFSZ;
exports.SSL_OP_ALL = SSL_OP_ALL;
exports.SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION;
exports.SSL_OP_CIPHER_SERVER_PREFERENCE = SSL_OP_CIPHER_SERVER_PREFERENCE;
exports.SSL_OP_CISCO_ANYCONNECT = SSL_OP_CISCO_ANYCONNECT;
exports.SSL_OP_COOKIE_EXCHANGE = SSL_OP_COOKIE_EXCHANGE;
exports.SSL_OP_CRYPTOPRO_TLSEXT_BUG = SSL_OP_CRYPTOPRO_TLSEXT_BUG;
exports.SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS = SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS;
exports.SSL_OP_EPHEMERAL_RSA = SSL_OP_EPHEMERAL_RSA;
exports.SSL_OP_LEGACY_SERVER_CONNECT = SSL_OP_LEGACY_SERVER_CONNECT;
exports.SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER = SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER;
exports.SSL_OP_MICROSOFT_SESS_ID_BUG = SSL_OP_MICROSOFT_SESS_ID_BUG;
exports.SSL_OP_MSIE_SSLV2_RSA_PADDING = SSL_OP_MSIE_SSLV2_RSA_PADDING;
exports.SSL_OP_NETSCAPE_CA_DN_BUG = SSL_OP_NETSCAPE_CA_DN_BUG;
exports.SSL_OP_NETSCAPE_CHALLENGE_BUG = SSL_OP_NETSCAPE_CHALLENGE_BUG;
exports.SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG = SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG;
exports.SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG = SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG;
exports.SSL_OP_NO_COMPRESSION = SSL_OP_NO_COMPRESSION;
exports.SSL_OP_NO_QUERY_MTU = SSL_OP_NO_QUERY_MTU;
exports.SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION;
exports.SSL_OP_NO_SSLv2 = SSL_OP_NO_SSLv2;
exports.SSL_OP_NO_SSLv3 = SSL_OP_NO_SSLv3;
exports.SSL_OP_NO_TICKET = SSL_OP_NO_TICKET;
exports.SSL_OP_NO_TLSv1 = SSL_OP_NO_TLSv1;
exports.SSL_OP_NO_TLSv1_1 = SSL_OP_NO_TLSv1_1;
exports.SSL_OP_NO_TLSv1_2 = SSL_OP_NO_TLSv1_2;
exports.SSL_OP_PKCS1_CHECK_1 = SSL_OP_PKCS1_CHECK_1;
exports.SSL_OP_PKCS1_CHECK_2 = SSL_OP_PKCS1_CHECK_2;
exports.SSL_OP_SINGLE_DH_USE = SSL_OP_SINGLE_DH_USE;
exports.SSL_OP_SINGLE_ECDH_USE = SSL_OP_SINGLE_ECDH_USE;
exports.SSL_OP_SSLEAY_080_CLIENT_DH_BUG = SSL_OP_SSLEAY_080_CLIENT_DH_BUG;
exports.SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG = SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG;
exports.SSL_OP_TLS_BLOCK_PADDING_BUG = SSL_OP_TLS_BLOCK_PADDING_BUG;
exports.SSL_OP_TLS_D5_BUG = SSL_OP_TLS_D5_BUG;
exports.SSL_OP_TLS_ROLLBACK_BUG = SSL_OP_TLS_ROLLBACK_BUG;
exports.S_IFBLK = S_IFBLK;
exports.S_IFCHR = S_IFCHR;
exports.S_IFDIR = S_IFDIR;
exports.S_IFIFO = S_IFIFO;
exports.S_IFLNK = S_IFLNK;
exports.S_IFMT = S_IFMT;
exports.S_IFREG = S_IFREG;
exports.S_IFSOCK = S_IFSOCK;
exports.S_IRGRP = S_IRGRP;
exports.S_IROTH = S_IROTH;
exports.S_IRUSR = S_IRUSR;
exports.S_IRWXG = S_IRWXG;
exports.S_IRWXO = S_IRWXO;
exports.S_IRWXU = S_IRWXU;
exports.S_IWGRP = S_IWGRP;
exports.S_IWOTH = S_IWOTH;
exports.S_IWUSR = S_IWUSR;
exports.S_IXGRP = S_IXGRP;
exports.S_IXOTH = S_IXOTH;
exports.S_IXUSR = S_IXUSR;
exports.UV_UDP_REUSEADDR = UV_UDP_REUSEADDR;
exports.W_OK = W_OK;
exports.X_OK = X_OK;
exports["default"] = api;

exports = module.exports = api;
